### DISCUSSION TITLE
cloud-custodian lambda with federated assumed-role in AWS

### DISCUSSION THREAD

**Discussion by dbdavisson** [cloud-custodian/cloud-custodian#9227] [AWS]
I am doing a POC for Cloud Custodian for our group and as I am going thru the use-cases I have ran into the below issue.

My AWS account is using Federated Web Identity using AWS IAM Identity Center.
I have a role named cloud-custodian-lambda-execution-role  
with 
AWSLambda_FullAccess  and AWSConfigRulesExecutionRole permission policies,  

and trust policy of

{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Service": "lambda.amazonaws.com"
            },
            "Action": "sts:AssumeRole"
        }
    ]
}

I have a policy statement below:
policies:

  - name: s3-tag-compliance-lambda
    resource: s3
    comment: |
      Report on s3 non tag compliant buckets
    mode:
      type: periodic
      schedule: rate(1 hour)
      role: arn:aws:iam::xxxxxxxxxxxx:role/cloud-custodian-lambda-execution-role
    description: s3 bucket without required tags
    filters:
      - or:
          - "tag:Owner": absent
          - "tag:CostCenter": absent
          - "tag:Project": absent
          
when I run any custodian policy that isn't trying to create a lambda it all works fine but when trying to create a lambda I get error, The role defined for the function cannot be assumed by Lambda.

From my own personal AWS account for root if I run the same policy with the same role it creates the Lambda just fine.
My question is this AWS IAM Identity Center a supported configuration like this and if so what do I need to add to the role to make it create the lambda function?  

Additionally I was able to configure c7n-mailer successfully on this account but it seems to run a bit differently.

Thanks for any help you can shed on this.

pip list
\Package                   Version
------------------------- ----------
argcomplete               3.1.6
attrs                     23.1.0
boto3                     1.33.1
botocore                  1.33.1
c7n                       0.9.33
c7n-mailer                0.6.32
certifi                   2023.11.17
charset-normalizer        3.3.2
datadog                   0.34.1
decorator                 5.1.1
docutils                  0.18.1
idna                      3.6
importlib-metadata        5.2.0
Jinja2                    3.1.2
jmespath                  1.0.1
jsonpatch                 1.33
jsonpointer               2.4
jsonschema                4.20.0
jsonschema-specifications 2023.11.1
ldap3                     2.9.1
MarkupSafe                2.1.3
pip                       22.0.2
pyasn1                    0.5.1
python-dateutil           2.8.2
python-http-client        3.3.7
PyYAML                    6.0.1
redis                     3.5.3
referencing               0.31.0
requests                  2.31.0
rpds-py                   0.13.1
s3transfer                0.8.0
sendgrid                  6.10.0
setuptools                59.6.0
six                       1.16.0
starkbank-ecdsa           2.2.0
tabulate                  0.9.0
types-six                 1.16.21.9
urllib3                   1.26.18
zipp                      3.17.0

**Comment by ajkerrigan** [comment]
Hmm that scenario should be fine, but if you're having trouble deploying those Lambda functions you might need to confirm that your credentials have `iam:PassRole` permissions on the Lambda execution role.

One thing you might try as a troubleshooting step is using the AWS CLI or web console to create/run a test Lambda function using the same execution role, taking Custodian out of the loop.

**Reply by dbdavisson** [reply]
Hi thanks for the feedback.  So I tried creating a lambda via the AWS CLI using the same role mentioned above and it created fine.  I used the basic hello world lambda example so the role seems to be fine from a lambda creation and execution perspective.

**Reply by ajkerrigan** [reply]
Ok that's good to know. I'm not sure what's dying though. You're seeing something like this when you run the policy? 

> botocore.errorfactory.InvalidParameterValueException: An error occurred (InvalidParameterValueException) when calling the UpdateFunctionConfiguration operation: The role defined for the function cannot be assumed by Lambda.

(Though I'd guess it'd be CreateFunction instead of UpdateFunctionConfiguration in your case)

While it sounds silly, it might be worth confirming that the role ARN in your policy exactly matches the working one you tested (if you haven't already). I mention this only because I've triggered that "role cannot be assumed" error on occasion with typos or flubbing a paste. Lambda says it can't assume the role in that case, but it's because the role doesn't exist.

It's possible to do some side-by-side troubleshooting, comparing the failing API calls Custodian makes against the successful ones you're doing manually. But that can get pretty tricky to talk through in an issue. Nicer to rule out as much as possible first!
